source/common/network/connection_impl.cc

```diff
@@ -474,6 +486,11 @@ void ConnectionImpl::updateWriteBufferStats(uint64_t num_written, uint64_t new_s
                                            connection_stats_->write_current_);
 }
 
+bool ConnectionImpl::isDoubleHalfClosed() {
+  // If the write_buffer_ is not empty, then the last_byte has not been sent to the transport yet.
+  return read_last_byte_ && write_last_byte_ && write_buffer_->length() == 0;
```

Comment:
Can we have an integration test for the following?

I want to make sure if we've read the last, and read the fin, and *not* propagated the last byte all the way through to the other end because the socket backed up, but the other side has half-closed, that the tcp proxy session sticks around until that data is flushed.  So basically half close one side, send GIANT data and immediately half close the other side.  I think if you don't drain the far end, you can force backing up and verify via watermark stats, then drain the far end and verify all the data made it through.  I think it's basically cloning the LargeWrite test but doing the large writes together and immediately half-closing, then draining.

I think this is functionally similar to the case one side closed and we verify we've flushed before we tear things down but good to have covered.